Date: Tue, 28 May 2002 21:24:25 -0400
From: Nathan Sharp <spamnps+jabber@phoenix-int.com>
To: jdev@jabber.org
Subject: [JDEV] Patch for server SSL hang in jabberd when client is slow authenticating

This patch fixes a bug where if a client connecting via SSL doesn't
authenticate very quickly, the server will hang until the client
finishes.  This hang requires a kill -9 to force the server to stop (in
the case where the client hangs while authenticating).  This happens,
for example, using Exodus when it displays one of the warning messages
about the certificate being invalid in some way (expired, wrong cn,
etc).   The server will stay hung until the client O.K.'s or cancels the
dialog.

The fix is to set the socket to non-blocking mode immediately after
accepting it, as non-blocking mode is not inherited from the listening
socket.  I'd appreciate it if this could make it into the next patch
release of 1.4 series!  This patch is against the 1.4.2 source code.

 Nathan

diff -durN jabber-1.4.2.orig/jabberd/mio_ssl.c jabber-1.4.2/jabberd/mio_ssl.c
--- jabber-1.4.2.orig/jabberd/mio_ssl.c	Fri Feb  8 08:39:27 2002
+++ jabber-1.4.2/jabberd/mio_ssl.c	Fri Jun  7 18:20:33 2002
@@ -219,6 +219,7 @@
     SSL_CTX *ctx = NULL;
     int fd;
     int sret;
+    int flags;
 
     if(m->ip == NULL)
     {
@@ -228,6 +229,12 @@
 
     fd = accept(m->fd, serv_addr, addrlen);
     
+    /* set the socket to non-blocking as this is not
+       inherited */
+    flags =  fcntl(fd, F_GETFL, 0);
+    flags |= O_NONBLOCK;
+    fcntl(fd, F_SETFL, flags);
+
     ctx = ghash_get(ssl__ctxs, m->ip);
     if(ctx == NULL)
     {
@@ -235,7 +242,8 @@
         return -1;
     }
     ssl = SSL_new(ctx);
-    log_debug(ZONE, "SSL accepting socket with new session %x", ssl);
+    log_debug(ZONE, "SSL accepting socket from %s with new session %x",
+                                   m->ip, ssl);
     SSL_set_fd(ssl, fd);
     SSL_set_accept_state(ssl);
     sret = SSL_accept(ssl);
